A68-deterministic-subsetting-lb-policy.md

```diff
@@ -0,0 +1,177 @@
+  repeated LoadBalancingConfig child_policy = 4;
+}
+```
+
+As you can see, the fields in this policy match exactly the fields in the deterministic subsetting LB service config.
+
+#### Integration with xDS LB Policy Registry
+As described in [gRFC A52][A52], gRPC has an LB policy registry, which maintains a list of converters. Every converter translates xDS LB policy to the corresponding service config. In order to allow using the Deterministic subsetting LB policy via xDS, the only thing that needs to be done is providing a corresponding converter function. The function implementation will be trivial as the fields in the xDS LB policy will match exactly the fields in the service config.
+
+## Rationale
+### Alternatives Considered: Deterministic Aperture
+Apart from Google's Deterministic Subsetting algorithm, we also considered [Twitter's Deterministic Aperture](https://patentimages.storage.googleapis.com/1a/09/5a/14392e76e5d8c5/US11119827.pdf) algorithm to handle subsetting. A critical difference in Twitter's algorithm is that it "splits" endpoints into multiple subsets. This is well demonstrated in their [ring diagrams](https://twitter.github.io/finagle/guide/ApertureLoadBalancers.html#deterministic-subsetting), where you can see a single service split into multiple subsets proportionately (i.e. 1/3 in one subset and 2/3 in the other). This ensures that even with simple child policies like `p2c`, the proportion of traffic sent to each backend across all subsets is balanced. 
+
+Though this balancing is valuable, we opted for Google's algorithm for several reasons:
+* Both algorithms are sufficient in reducing connection count, which is the primary drawback of `round_robin`. 
+* Both algorithms provide better load distribution guarantees than `pick_first`. 
+* When the client to backend ratio requires overlapping subsets, deterministic subsetting provides the additional guarantee of unique subsets. Because the aperture algorithm is built around a fixed ring, when clients \* aperture exceeds the number of servers, aperture takes "laps" over the ring, producing similar subsets for several clients. Deterministic subsetting strives to reduce the risk of misbehaving endpoints by distributing them more randomly across client subsets.  
+* In order to take advantage of the partial weighting of backends, Twitter's algorithm would require passing weight information from parent balancer to child balancer. The appropriate balancer to handle this sort of weighting would be weighted\_round\_robin, but weighted\_round\_robin currently calculates weights based on server load alone, and it is not designed to consider this additional weighting information. We opted for the solution that allowed us to keep this existing balancer as is. Though deterministic subsetting does not guarantee even load distribution across subsets, its diverse subset paired with weighted\_round\_robin as the child policy within subsets should be sufficient for most use cases.  
```

Comment:
It's true that the gRPC WRR policy described in [gRFC A58](https://github.com/grpc/proposal/blob/master/A58-client-side-weighted-round-robin-lb-policy.md) uses weight information provided by the endpoints.  However, there is a separate community-contributed proposal open (although admittedly hasn't been active for a while) to provide an Envoy-style WRR policy that uses weights specified by the control plane in [pending gRFC A34](https://github.com/grpc/proposal/pull/202).  The two policies are not mutually exclusive -- we can support both if someone wants to contribute the latter.

gRPC C-core does [pass endpoint weights down from the xDS control plane](https://github.com/grpc/grpc/blob/de98c1c9ad322a680a844f14f93166403dad6650/src/core/ext/xds/xds_endpoint.cc#L226) to the LB policy, but it's up to the individual LB policy as to whether it uses that information.  The only policy we currently have that does so is the [ring_hash policy](https://github.com/grpc/grpc/blob/de98c1c9ad322a680a844f14f93166403dad6650/src/core/ext/filters/client_channel/lb_policy/ring_hash/ring_hash.cc#L474).  I believe gRPC Go passes the weight through as well, but I don't know about Java.  And even if they didn't, I don't think it would be hard to add this plumbing for any gRPC implementation that supports xDS.

My reason for mentioning all of this is just to say that the need for endpoint weights from the control plane is not necessarily a reason not to support deterministic aperature subsetting in the general case.  But it would make sense to say that deterministic aperature would not work with a control plane that does not send weight information or with an LB policy like gRPC's WRR that uses out-of-band weight information.